crates/sui-graphql-rpc/src/types/dot_move/data_loader.rs

```diff
@@ -0,0 +1,191 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::{collections::HashMap, str::FromStr, sync::Arc};
+
+use async_graphql::dataloader::{DataLoader, Loader};
+use serde::{Deserialize, Serialize};
+
+use crate::{error::Error, types::base64::Base64};
+
+use super::config::{AppRecord, DotMoveConfig, DotMoveServiceError, Name};
+
+/// GraphQL fragment to query the values of the dynamic fields.
+const QUERY_FRAGMENT: &str =
+    "fragment RECORD_VALUES on DynamicField { value { ... on MoveValue { bcs } } }";
+
+pub(crate) struct MainnetNamesLoader {
+    client: reqwest::Client,
+    config: DotMoveConfig,
+}
+/// Helper types for accessing a shared `DataLoader` instance.
+#[derive(Clone)]
+pub(crate) struct DotMoveDataLoader(pub Arc<DataLoader<MainnetNamesLoader>>);
+
+impl MainnetNamesLoader {
+    pub(crate) fn new(config: &DotMoveConfig) -> Self {
+        Self {
+            client: reqwest::Client::new(),
+            config: config.clone(),
+        }
+    }
+
+    /// Constructs the GraphQL Query to query the names on a mainnet graphql endpoint.
+    pub(crate) fn construct_names_graphql_query(
+        &self,
+        names: &[Name],
+        mapping: &mut HashMap<Name, usize>,
+    ) -> String {
+        let mut result = format!(r#"{{ owner(address: "{}") {{"#, self.config.registry_id);
+
+        // we create the GraphQL query keys with a `fetch_{id}` prefix, which is accepted on graphql fields.
+        for (index, name) in names.iter().enumerate() {
+            let bcs_base64 = name.to_base64_string();
+
+            // retain the mapping here (id to bcs representation, so we can pick the right response later on)
+            mapping.insert(name.clone(), index);
+
+            let field_str = format!(
+                r#"{}: dynamicField(name: {{ type: "{}::name::Name", bcs: {} }}) {{ ...RECORD_VALUES }}"#,
+                fetch_key(&index),
+                self.config.package_address,
+                bcs_base64
+            );
+
+            result.push_str(&field_str);
```

Comment:
We're writing to a String, not a buffer, so the `write` macro doesn't seem to do what we want.
assignment is not needed anyways, I just did an immediate `result.push_str(&format!()` to avoid the assignment instead.